### DISCUSSION TITLE
Fields "mandatories dynamically"

### DISCUSSION THREAD

**Discussion by rengit62** [OP-TED/eForms-SDK#485] [Q&A] [answered]
Hi Lab !
In a situation like this : 
![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/beb20591-d377-4c2d-baf9-d0cdd123c15f)

the xml is wrong because it lacks OPT-300-Procedure-SProvider.
And there are many of these situations.
How do you manage this problem with sdk ?
Of course I could write a simple javascript function, but this would be opposite to metadata-driven applications concept
Bye
Renato

**Comment by rouschr** [comment, marked as answer]
We call these "dynamic rules" or "conditional rules", ...
For example for OPT-300-Procedure-SProvider the condition (in fields.json) is:
  ```
  ...
    "assert" : {
      "value" : "{ND-Root} ${TRUE}",
      "severity" : "ERROR",
      "constraints" : [ {
        "condition" : "{ND-ServiceProvider} ${OPT-300-Procedure-SProvider is present}",
        "value" : "{ND-ServiceProvider} ${every text:$provider in OPT-300-Procedure-SProvider satisfies ($provider in /OPT-200-Organization-Company)}",
        "severity" : "ERROR",
        "message" : "rule|text|BR-OPT-00300-0251"
      } ]
    }
```
This is in the EFX language, thus this is part of the meta data as it is part of fields.json
EFX is a language specific to the SDK.

      
For dynamic rules you can use EFX. We have plans to demo this later in the editor demo (either fully using JS or on server side), but you will have to be patient for that because many other things will take priority.
EFX rules can be evaluated directly in the browser using javascript (not very easy ...) or on the server side using Java/C#... & XML/XPATH (easier). We provide a Java implementation to evaluate EFX: https://github.com/OP-TED/efx-toolkit-java
If you do not use Java on the server side you could also just have a little service running Java and call it from your server, that would be the simplest option.
There are also features foreseen for giving better feedback in the UI using only CVS (parsing svrl, and reading a custom report on the client side).

**Reply by rengit62** [reply]
Hi ! I cloned the project toolkit and made a little main to understand how it works ..
Just a question ..
With EFX you can make an EFX espression (following ELX rules, of course) , and the java classes convert this expression in a xpath string ?
After, you can compile and evaluate the xpath expression, checking the espression , that is the condition ?
Thanks and Regards
Renato

**Reply by rouschr** [reply]
Yes the provided Java EFX implementation relies on xpath expressions.
See docs for details: https://docs.ted.europa.eu/eforms/latest/efx-toolkit/
The EFX ANTLR grammar is part of the SDK: https://github.com/OP-TED/eForms-SDK/tree/develop/efx-grammar
The efx toolkit supports multiple SDK versions (see code like @SdkComponent(versions = {"0.6", "0.7", "1"},
    componentType = SdkComponentType.SYMBOL_RESOLVER)

>  After, you can compile and evaluate the xpath expression, checking the espression , that is the condition ?

Yes for rules the conditions mostly evaluate to boolean true/false (fields.json).

I think for now you are interested in the rules that evaluate to either true or false that you can find in the fields.json, more advanced rules can do more for example in the future some rules may go as far as looking up values in other existing notices (URL lookup).

I recommend you look into the unit tests:
https://github.com/OP-TED/efx-toolkit-java/blob/develop/src/test/java/eu/europa/ted/efx/EfxExpressionTranslatorTest.java

**Reply by rengit62** [reply]
Thanks, very interesting.
for the data, I saw that DependencyFactoryMock create SymbolResolverMock which has got the mock data ( public void loadMapData() throws JsonMappingException, JsonProcessingException {

For a 'real test' should I load the fields of fields.json instead ?
Regards
Renato

**Reply by rouschr** [reply]
See the real resolver here: https://github.com/OP-TED/efx-toolkit-java/blob/develop/src/main/java/eu/europa/ted/eforms/sdk/SdkSymbolResolver.java#L61

```
  protected Map<String, SdkField> fieldById;
  protected Map<String, SdkNode> nodeById;

 protected void loadMapData(final String sdkVersion, final Path sdkRootPath)
      throws InstantiationException {
    Path jsonPath = SdkResourceLoader.getResourceAsPath(sdkVersion,
        SdkConstants.SdkResource.FIELDS_JSON, sdkRootPath);
...
```

**Reply by rengit62** [reply]
Thanks !
Ok for the SdkSymbolResolver instead of the SymbolResolverMock , but I found only one implementation for TranslatorDependencyFactory

![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/aff207e9-7e19-4b35-9dd6-76720d8aafbd)

Am I missing something ? 
Bye !
Renato

**Reply by rouschr** [reply]
You can find a dependency factory implementation in the viewer project: https://github.com/OP-TED/eforms-notice-viewer

```
package eu.europa.ted.eforms.viewer;
...
public class DependencyFactory implements TranslatorDependencyFactory {
```

**Reply by elerik** [reply]
Hi 
i wrote that "simple javascript" which dynamically solve it in real time (for most of it)
during rendering i insert attribute for mandatory and forbidden conditions if they have some (assert i dont implement at this moment) and tag them as conditional-required.
Then i go through only elements with those attributes and solve them. Even with complex form its surprisingly not so much elements so its pretty fast. 
Then i evaluate efx condition via javascript and adapt element requirements It wasnt so simple how i thought it will be.
for example most of elements are just part of groups so i evaluate firstly only elements in same group. then if there is none (in some cases) i expand this area of check to parent group and so on until its whole form (rare situations) 
I have never met efx so i evaluate those "string conditions" in my own. There are i guess some libraries for that. 

regards

**Reply by rengit62** [reply]
I am trying to put EFX conversion inside app ( application server wildfly ).
I copy  DependencyFactory from https://github.com/OP-TED/eforms-notice-viewer, in project efx-toolkit-java .
I only commented (SdkDownloader.downloadSdk(sdkVersion, sdkRoot);) because of an error on MavenBooter 

But here : 

![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/07399f1d-4351-45a7-98d0-570db14e1dfa)


I got an error : 
![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/13fe25d8-b58a-44a3-91f6-eb89cede3a78)


can you help me ?
Thanks and Regards




,

**Reply by rouschr** [reply]
I am not sure what your error is as I see no stack trace. Looks like it does not find something for the `SCRIPT_GENERATOR` as indicated by your screenshot but other things only show "Object...." so I cannot help you. A Java error logs with a stacktrace would be more helpful.
In the worst case you could manually `git clone` version 1.8.0 from the GitHub SDK project and put it into the sdks folder yourself if the problem is related to downloading from GitHub.
Also make sure to use the latest version of the editor demo project which now supports 1.8 "xsdSequenceOrder", later on it may only support 1.9, so always make sure you check the supported versions in `application.yaml` of the editor. Latest work on EFX toolkit may rely on data or features of 1.9, so make sure you have no mix of 1.8 and 1.9 between the used projects.

**Comment by rengit62** [comment]
Thanks Christophe
I've just got to a result, changing some classes, statements, but I would like to "consolidate" the result, without changing anything.

Some questions to begin : 
1) you told me to get DependencyFactory class from https://github.com/OP-TED/eforms-notice-viewer, I put here in project efx-toolkit-java, it is right ?
![image](https://github.com/OP-TED/eForms-SDK/assets/52211037/136932d3-6d84-4725-8718-06161b2ce423)
Will this class be put in the future in toolkit/core ?

2) compiling with maven I set 2 dependencies in our current project  : 

`
<dependency>
    <groupId>eu.europa.ted.eforms</groupId>
    <artifactId>efx-toolkit-java</artifactId>
    <version>${efx-toolkit.version}</version>
 </dependency>
< dependency>
    <groupId>eu.europa.ted.eforms</groupId>
    <artifactId>eforms-core-java</artifactId>
    <version>${eforms-core.version}</version>
  </dependency>
`

Do you agree with level of project ? 

About 1.8 I've already downloaded  and I am using it.


3) let's consider for example : 
  "condition" :  {ND-ContractingParty} ${BT-11-Procedure-Buyer not in ('body-pl','body-pl-cga','body-pl-la','body-pl-ra','cga','def-cont','eu-ins-bod-ag','grp-p-aut','int-org','la','org-sub','org-sub-cga','org-sub-la','org-sub-ra','ra')}
      
about context , of course I can split the string to find ND-ContractingParty and make the context, is there any method , or class that is present in toolkit/core to get the context, or at list the first string (ND-ContractingParty) ?


Bye
Renato

**Comment by rouschr** [comment]
Hello, 
I never said to get or copy Java files somewhere, you can see how this works by looking at EFX toolkit or the viewer project. From there on you have to use your Java knowledge.
For EFX conditions, you should use the EFX toolkit.
For EFX templates, you should use the viewer (see `view-templates` .efx files of the SDK).
For both the core project provides some classes.
In the toolkit project we extend the `SdkComponentFactory` of the core project, I think in previous answers I forgot to mention the core project: https://github.com/OP-TED/eforms-core-java/blob/develop/src/main/java/eu/europa/ted/eforms/sdk/component/SdkComponentFactory.java
I recommend that in general you look into junit tests (look for `@Test`) of the projects as this shows you how the code can be run to evaluate expressions, for example in EFX toolkit see:
```
class EfxExpressionCombinedV1Test extends EfxTestsBase {
  @Override
  protected String getSdkVersion() {
    return "eforms-sdk-1.0";
  }

  @Test
  void testNotAnd() {
    testExpressionTranslationWithContext("not(1 = 2) and (2 = 2)", "BT-00-Text",
        "not(1 == 2) and (2 == 2)");
  }

  @Test
  void testNotPresentAndNotPresent() {
    testExpressionTranslationWithContext("not(PathNode/TextField) and not(PathNode/IntegerField)",
        "ND-Root", "BT-00-Text is not present and BT-00-Integer is not present");
  }
  ...
```

**Reply by rengit62** [reply]
Hi
for now we are interested to expression, only.

in https://github.com/OP-TED/efx-toolkit-java
it is written : 
**The EFX toolkit requires Java 11 or later.

It is available as a Maven package on Maven Central and can be used by adding the following to the project's pom.xml.

<dependencies>
  ...
`<dependency>
    <groupId>eu.europa.ted.eforms</groupId>
    <artifactId>efx-toolkit-java</artifactId>
    <version>${efx-toolkit.version}</version>
  </dependency>
`
    ...
</dependencies>**

and nothing else.

As the junit tests you are written about, use the class DependencyFactoryMock, are you going to put on toolkit an implementation of TranslatorDependencyFactory that use the real fields.json and codelist ?
Best Regards

**Reply by rouschr** [reply]
Hello, I do not think we will do that in the toolkit, but as said in another answer, something similar is done in the viewer, there the SDK is used. For SDK download automation see usage of https://github.com/OP-TED/eforms-core-java/blob/develop/src/main/java/eu/europa/ted/eforms/sdk/resource/SdkDownloader.java
You can also just download SDK files manually and put them in a folder for a start, but we recommend you use the SDK downloader. The viewer uses it.
